### ISSUE TITLE
BatchWriteItem should check whether there are more than 25 requests in the batch

### ISSUE THREAD

**Issue by Rjerk** [scylladb/scylladb#9389]
As the [DynamoDB document](https://docs.aws.amazon.com/amazondynamodb/latest/APIReference/API_BatchWriteItem.html) says:

> If one or more of the following is true, DynamoDB rejects the entire batch write operation:
> ...
> There are more than 25 requests in the batch.

AWS Dynamodb reports errors like:

> An error occurred (ValidationException) when calling the BatchWriteItem operation: 1 validation error detected: Value [...] at 'requestItems' failed to satisfy constraint: Map value must satisfy constraint: [Member must have length less than or equal to 25, Member must have length greater than or equal to 1]

But the Alternator does not.

**Comment by psarna** [comment]
I just noticed that you (correctly) created an issue first and only then sent a pull request, so I'll just link my response here: https://github.com/scylladb/scylla/pull/9390#issuecomment-927602379, since it's actually relevant for the issue.

**Comment by nyh** [comment]
This issue is a duplicate of issue #5057, although issue #5057 (perhaps unwisely) combined a few BatchWriteItem limits and not just the "25".

I have test titled "[PATCH] test/alternator: add test for too many items in BatchWriteItem" waiting for review on the mailing list (hint, hint) which confirms that this limit 25 exists on DynamoDB but not on Scylla.

I agree with @psarna that we could decide to allow more than 25 items in a batch. We hopefully shouldn't have the concurrency problems of reads (#5944) in the writing side. But we need to confirm that we don't. We also have some smaller problems remaining with large batches (#8183).

**Comment by nyh** [comment]
> We hopefully shouldn't have the concurrency problems of reads (#5944) in the writing side.

On second thought, we do have a concurrency problem with writes as well. Because we (usually/often) use LWT for writes, writes to different partitions involve different locking, and can and should be done in parallel, so it does make sense to do many writes in parallel, but if a batch has 1000 writes it probably doesn't make sense to use parallelism of 1000.

**Comment by gleb-cloudius** [comment]
CQL batch allows LWT write to only one partition. If Alternator's batch writes use CQL batch writes then only one partition can be modified at a time.

**Comment by Rjerk** [comment]
Make sense. Probably a more considerable value is a better idea, not just 25.

**Comment by nyh** [comment]
> CQL batch allows LWT write to only one partition. If Alternator's batch writes use CQL batch writes then only one partition can be modified at a time.

DynamoDB's notion of a "batch" is unrelated to Cassandra's they just name the same name, unfortunately: A DynamoDB "batch" is just a set of requests that will be done in parallel without any special isolation guarantees (it's not a transaction in any sense of the word). DynamoDB's batch is needed because the HTTP 1.* transport used by DynamoDB doesn't have support for multiple request streams, so these "batches" allow you to achieve high concurrency with fewer (25 times fewer) connections - I discussed this in my blog post comparing the CQL and DynamoDB APIs, https://www.scylladb.com/2020/05/12/comparing-cql-and-the-dynamodb-api/

Arguably, we could support "batches" (in the DynamoDB, not Cassandra, sense) of 1000 requests, but internally limit their concurrency to something smaller (even that number 25). But a simpler alternative is **not** to allow more than 25 items in a batch, and ask the client to split the batches. DynamoDB clients already know that no more than 25 items are allowed, and split large batches into smaller batches of 25 - e.g., the Python client does this for you - see https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/dynamodb.html#DynamoDB.Table.batch_writer - so we're not really losing much if we implement the same limit 25 as DynamoDB does.

Even more importantly, if we do choose a different limit - e.g., 50 - clients would have no convenient way to inquire what the new limit is. Existing clients will continue to split batches to 25 in each batch (because that's DynamoDB's limit), and which want to figure out the new limit is 50 will need to do something ugly, like use the low-level BatchWriteItem function and increase the number of items in each attempt until getting an error that there are two many items in the batch.

**Comment by nyh** [comment]
Closing as a duplicate of #5057. Added a few insights from here to that issue.
